rviz_common/src/rviz_common/depth_cloud_mld.cpp

```diff
@@ -0,0 +1,601 @@
+
+    // precompute 3D projection matrix
+    for (uint32_t v = 0; v < height; ++v, ++projY) {
+      *projY = (v - center_y) * constant_y;
+    }
+
+    for (uint32_t u = 0; u < width; ++u, ++projX) {
+      *projX = (u - center_x) * constant_x;
+    }
+
+    // reset shadow vectors
+    reset();
+  }
+}
+
+
+template<typename T>
+sensor_msgs::msg::PointCloud2::SharedPtr
+MultiLayerDepth::generatePointCloudSL(
+  const sensor_msgs::msg::Image::ConstSharedPtr & depth_msg,
+  std::vector<uint32_t> & rgba_color_raw)
+{
+  int width = depth_msg->width;
+  int height = depth_msg->height;
+
+  sensor_msgs::msg::PointCloud2::SharedPtr cloud_msg = initPointCloud();
+  cloud_msg->data.resize(height * width * cloud_msg->point_step);
+
+  uint32_t * color_img_ptr = nullptr;
+
+  if (!rgba_color_raw.empty()) {
+    color_img_ptr = &rgba_color_raw[0];
+  }
+
+  ////////////////////////////////////////////////
+  // depth map to point cloud conversion
+  ////////////////////////////////////////////////
+
+  float * cloud_data_ptr = reinterpret_cast<float *>(&cloud_msg->data[0]);
+
+  std::size_t point_count = 0;
+  std::size_t point_idx = 0;
+
+  const T * depth_img_ptr = reinterpret_cast<const T *>(&depth_msg->data[0]);
+
+  std::vector<float>::iterator proj_x;
+  std::vector<float>::const_iterator proj_x_end = projection_map_x_.end();
+
+  std::vector<float>::iterator proj_y;
+  std::vector<float>::const_iterator proj_y_end = projection_map_y_.end();
+
+  // iterate over projection matrix
+  for (proj_y = projection_map_y_.begin(); proj_y != proj_y_end; ++proj_y) {
+    for (proj_x = projection_map_x_.begin(); proj_x != proj_x_end;
+      ++proj_x, ++point_idx, ++depth_img_ptr)
+    {
+      T depth_raw = *depth_img_ptr;
+      if (DepthTraits<T>::valid(depth_raw)) {
+        float depth = DepthTraits<T>::toMeters(depth_raw);
+
+        // define point color
+        uint32_t color;
+        if (color_img_ptr) {
+          color = *color_img_ptr;
+        } else {
+          color = ((uint32_t)255 << 16 | (uint32_t)255 << 8 | (uint32_t)255);
+        }
+
+        // fill in X,Y,Z and color
+        *cloud_data_ptr = (*proj_x) * depth;
+        ++cloud_data_ptr;
+        *cloud_data_ptr = (*proj_y) * depth;
+        ++cloud_data_ptr;
+        *cloud_data_ptr = depth;
+        ++cloud_data_ptr;
+        *cloud_data_ptr = *reinterpret_cast<float *>(&color);
+        ++cloud_data_ptr;
+
+        ++point_count;
+      }
+
+      // increase color iterator pointer
+      if (color_img_ptr) {
+        ++color_img_ptr;
+      }
+    }
+  }
+
+  finalizePointCloud(cloud_msg, point_count);
+
+  return cloud_msg;
+}
+
+
+template<typename T>
+sensor_msgs::msg::PointCloud2::SharedPtr
+MultiLayerDepth::generatePointCloudML(
+  const sensor_msgs::msg::Image::ConstSharedPtr & depth_msg,
+  std::vector<uint32_t> & rgba_color_raw,
+  ros_integration::RosNodeAbstractionIface::WeakPtr rviz_ros_node)
+{
+  int width = depth_msg->width;
+  int height = depth_msg->height;
```

Comment:
```suggestion
  uint32_t width = depth_msg->width;
  uint32_t height = depth_msg->height;
```

(actually, you could just drop these and directly use depth_msg->width and depth_msg->height in the single place they are used below)